### DISCUSSION TITLE
Barracuda_libre-12.2134.0 (Final Official Firmware)

### DISCUSSION THREAD

**Discussion by coggy9** [coggy9/HKHacking#11] [General]
This is the final OTA update that Harman produced for the Invoke. It was silently published to the Invoke's product page [here](https://support.harmankardon.com/us/en/speakers/speakers-bluetooth-wireless/INVOKE.html?color=graphite) as a FAQ question on September 8, 2021. 

It is mirrored here for archival purposes. Instructions are included as a PDF inside the ZIP file.

As for modifying this one, it's going to be a bit tougher, as the system partition went from over 100MB to just 66MB, leaving less room for files. You should be able to inject it into the old Cortana enabled 83_IMAGE, but other partitions in 83_IMAGE also have changes, so it may not boot/work correctly.  Every other file/image is identical to the first flashing package though.

There are also new binary files, such as `wifi-blocker`, which I assumes blocks WiFi, making the old ADB trick require more modifications. 

As expected, files relating to Cortana/Spotify are removed. 

`/usr/lib/gstreamer-1.0` is missing several files, this image may have been built with less codec support, since its expected to only play Bluetooth audio now, instead of streaming online content.

A new Bluetooth firmware is included at `/lib/firmware/mrvl/sd8887_bt_a2_new.bin`, not sure if it is used. 

These are just my casual 3AM observation into the image, so there are probably more changes that will need to be addressed before this image can be used for development.

<hr /><em>This discussion was created from the release <a href='https://github.com/coggy9/HKHacking/releases/tag/FinalOTA'>Barracuda_libre-12.2134.0 (Final Official Firmware)</a>.</em>

**Comment by avivhu** [comment]
My device didn't get the convert to stupid Bluetooth update and I thought it will never work again. Followed the instructions - it works!

**Comment by jim-liu** [comment]
One of my Invoke probably encountered some error when updating to the final firmware. It would stop accepting Bluetooth connection after a day or two. I followed the instructions and reflashed the firmware, now it works perfectly (well, as perfect as a Bluetooth speaker can be). Thanks!

**Comment by ductaitran** [comment]
Hi, hope you're still there, I just wonder after updating the firmware can I use my Invoke microphone for phone call? I think it has the necessary hardware to make calls (mics, speakers and Bluetooth connection), but seems the mic is disable after I update it to OTA2

**Reply by coggy9** [reply]
Downgrading to a Cortana enabled firmware should work, but I haven't tested it.

**Comment by baccccccc** [comment]
Does anyone have more precise instructions? I'm trying to follow the attached PDF but getting confused. It's either me doing something stupid, or my Invoke is completely broken.

It always boots to _"Hi there, I'm Cortana, I'm here to help. Let's get your Harman Kardon Invoke ready to use..."_ no matter what I'm doing.

(And even once booted normally, it cannot be located by the "Cortana Device Setup" app either on PC or my old Windows Phone. So maybe there's indeed a problem.)

Here's what I'm doing.

> Process 1. Installing Marvell USB driver" (from the PDF)

This part was completed successfully, to the best of my understanding. The USB device gets recognized by the driver, without any bangs or warnings.

<img width="315" alt="image" src="https://github.com/coggy9/HKHacking/assets/107458411/f8d948f4-ec89-4d92-9b5b-20321196b0c4">

...although it keeps disconnecting. (More on that later.)

> Process 2. Installing the final OTA2 software on the INVOKE

This is where I'm starting to get confused

> 1. Disconnect the INVOKE from electricity. 
> 2. Open the 'OTA2' folder 
> 3. Launch run.bat - console is opened.  

Done, and in fact I see an empty PuTTY window. Should it remain empty at this step?

> 4. Connect speaker with _ORDINARY_ USB cable to your PC. 
> 5. Using a paper clip or similar device, press and hold the Reset pinhole and plug the speaker back 
into electricity. 
> 6. While still holding the Reset pinhole, press the Mic Off button 4 (four) times on the speaker 
within 5 seconds.  
> • Note: press button 4 times exactly! Within a few seconds the speaker top panel should 
display yellow light momentarily to indicate flashing mode is On.  If it does not, repeat 
steps 3-6 until you succeed. 

I believe I'm getting this done, too. At least, I do see the yellow light for a few seconds.

> 7. Wait till U-boot screen appears in console, then release Reset pinhole. 

Now, I have no idea what the "U-boot screen" should look like. Perhaps a screenshot or a text snipped could help? 

But this is what I'm seeing in the previously empty PuTTY window.

```
one target device connected.
11832049b8f40a32i*m*g*r*qtarget device disconnected.
one target device connected.
11832049b8f40a32i*m*g*r*qtarget device disconnected.
one target device connected.
11832049b8f40a32i*m*g*r*qtarget device disconnected.
```

I've also tried to understand the following note but apparently it does not help.

> • If you see something like device disconnected instead of this screen for more than 10 
seconds, then unplug speaker, wait 10 seconds and plugin it again but do not touch USB, 
leave it connected. 

As I understand, it meant to say, "unplug and reconnect _power_ while keeping USB on" (as in the main procedure) but without pressing the microphone button four times. I tried that but it does not seem to make any difference. I just see more "connected" messages in PuTTY window, following immediately by "disconnected".

What may also indicate a problem is that the USB device keeps disappearing from the Device manager in Windows. Originally when the device is booted, it gets recognized briefly. Then it gets disconnected and disappears from the device manager in a few seconds. And then it flips on and off for a few times, and never comes back. And then, in a minute or so, the device boots to "Hi there, I'm Cortana" as if I have not done anything.

This correlates with what PuTTY shows, I guess. But once again, it might indicate that either I'm doing something wrong, or my Invoke is broken beyond repair.

**Reply by coggy9** [reply]
I never had any luck getting it to work on Windows, I would try using a Linux live CD/USB and flash your Invoke that way, IIRC the PDF also has instructions for Linux flashing.

**Reply by baccccccc** [reply]
hmm. Thanks. No, I don't see any Linux instructions in the PDF, but I'll keep looking. I'll probably try another Windows PC before that, but that's for another day. Happy holidays, anyway, and thanks for keeping the lights on here.

**Reply by rbeesley** [reply]
I _think_ I got it working via WSL2, but I'm not sure how I'd have connected to the USB port. I don't have a Linux distro installed, so it must have been something like that. I was using PuTTY though, so maybe part of it was via WSL and the other part was in Windows via the IP address.

**Reply by baccccccc** [reply]
Thanks, that helps. Although I'm now positive my Invoke somehow got damaged :( The problem is, the USB device gets disconnected a moment after it appears. I tried it on a few Windows machines (with and without Marvell driver installed) and it behaves the same. I watched a few YouTube videos to make myself more familiar with the process, and it looks like with a healthy speaker, it does not disconnect. (Or disconnects momentarily, and then comes back and stays.) In my case, it disappears completely.

I also saw some notes that if USB cable has insufficient strength (?), then the device won't connect at all. I tried a few cables I had laying around, the result is always the same. I might try to find a thicker cable before ruling out all possibilities.

**Reply by coggy9** [reply]
That's the same issue I had with mine. It would disconnect and not come back no matter what I tried. I'll try to get some quick Linux flashing instructions up sometime next ~~year~~ week

**Reply by coggy9** [reply]
@baccccccc  The steps for using Linux are mostly the same. I tested on a Manjaro live USB since I could not get a Ubuntu live USB to boot for some reason, but this should work for any distro.

1. Install PuTTY if it is not already installed
3. Extract the update ZIP
4. Open a terminal window in the OTA folder
5. Run `chmod a+x run.sh` and `chmod a+x usb_boot` to mark the files as executable
6. Run `sudo ./run.sh`
7. Get the Invoke into flashing mode, following the official guide PDF.
8. Once in the U-Boot terminal, run `l2nand 83`
9. After you see `Congratulations! U2nand succeed!`, unplug the Invoke from USB and power and plug it back in to power boot it normally.
